homeassistant/components/media_player/enigma.py

```diff
@@ -0,0 +1,250 @@
+from homeassistant.util import Throttle
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['beautifulsoup4==4.5.3']
+
+# Return cached results if last scan was less then this time ago.
+MIN_TIME_BETWEEN_SCANS = timedelta(seconds=10)
+
+DEFAULT_NAME = 'Enigma2 Satelite'
+DEFAULT_PORT = 80
+DEFAULT_TIMEOUT = None
+DEFAULT_USERNAME = 'root'
+DEFAULT_PASSWORD = 'password'
+
+SUPPORT_ENIGMA = SUPPORT_VOLUME_SET | SUPPORT_VOLUME_MUTE | \
+                 SUPPORT_TURN_ON | SUPPORT_TURN_OFF | \
+                 SUPPORT_SELECT_SOURCE
+
+MAX_VOLUME = 100
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_USERNAME, default=DEFAULT_USERNAME): cv.string,
+    vol.Optional(CONF_PASSWORD, default=DEFAULT_PASSWORD): cv.string,
+    vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+    vol.Optional(CONF_TIMEOUT, default=DEFAULT_TIMEOUT): cv.socket_timeout,
+})
+
+
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Setup the Enigma platform."""
+    enigma = EnigmaDevice(config.get(CONF_NAME),
+                          config.get(CONF_HOST),
+                          config.get(CONF_PORT),
+                          config.get(CONF_USERNAME),
+                          config.get(CONF_PASSWORD),
+                          config.get(CONF_TIMEOUT))
+    if enigma.update():
+        async_add_devices([enigma])
+        return True
+    return False
+
+
+class EnigmaDevice(MediaPlayerDevice):
+    """Representation of a Enigma device."""
+
+    def __init__(self, name, host, port, username, password, timeout):
+        """Initialize the Enigma device."""
+        self._name = name
+        self._host = host
+        self._port = port
+        self._username = username
+        self._password = password
+        self._timeout = timeout
+        self._pwstate = True
+        self._volume = 0
+        self._muted = False
+        self._selected_source = ''
+        self._source_names = {}
+        self._sources = {}
+        self.handle_base_auth()
+        self.load_sources()
+
+    def handle_base_auth(self):
+        """Handle HTTP Auth."""
+        mgr = urllib.request.HTTPPasswordMgrWithDefaultRealm()
+        mgr.add_password(None, self._host, self._username, self._password)
+        handler = urllib.request.HTTPBasicAuthHandler(mgr)
+        opener = urllib.request.build_opener(handler)
+        urllib.request.install_opener(opener)
+
+    def load_sources(self):
+        """Load sources from first bouquet."""
+        from bs4 import BeautifulSoup
+        reference = urllib.parse.quote_plus(self.get_bouquet_reference())
+        try:
+            epgbouquet_xml = self.request_call('/web/epgnow?bRef=' + reference)
+        except (HTTPError, URLError, ConnectionRefusedError):
+            return False
+
+        soup = BeautifulSoup(epgbouquet_xml, 'html.parser')
```

Comment:
There is too much protocol specific information in this integration. Please extract it into a standalone library that this integration can depend on .